realm/realm-library/src/main/java/io/realm/HandlerController.java

```diff
@@ -345,16 +408,78 @@ private boolean threadContainsAsyncQueries () {
         return !isEmpty;
     }
 
+    boolean threadContainsAsyncEmptyRealmObject() {
+        boolean isEmpty = true;
+        Iterator<Map.Entry<WeakReference<RealmObject>, RealmQuery<?>>> iterator = emptyAsyncRealmObject.entrySet().iterator();
+        while (iterator.hasNext()) {
+            Map.Entry<WeakReference<RealmObject>, RealmQuery<?>> next = iterator.next();
+            if (next.getKey().get() == null) {
+                // clean the GC'ed instances
+                // we could've avoided this if we had a 'WeakIdentityHashmap' data structure. miss Guava :(
+                iterator.remove();
+            } else {
+                isEmpty = false;
+            }
+        }
+
+        return !isEmpty;
+    }
+
     private void deleteWeakReferences() {
         // From the AOSP FinalizationTest:
         // https://android.googlesource.com/platform/libcore/+/master/support/src/test/java/libcore/
         // java/lang/ref/FinalizationTester.java
-        // System.gc() does not garbage collect every time. Runtime.gc() is
-        // more likely to perform a gc.
+        // System.gc() does not garbage collect every time. Runtime.gc() is more likely to perform a gc.
         Runtime.getRuntime().gc();
-        Reference<? extends RealmResults<? extends RealmObject>> weakReference;
-        while ((weakReference = referenceQueue.poll()) != null ) { // Does not wait for a reference to become available.
-            asyncRealmResults.remove(weakReference);
+        Reference<? extends RealmResults<? extends RealmObject>> weakReferenceResults;
+        Reference<? extends RealmObject> weakReferenceObject;
+        while ((weakReferenceResults = referenceQueueAsyncRealmResults.poll()) != null ) { // Does not wait for a reference to become available.
+            asyncRealmResults.remove(weakReferenceResults);
+        }
+        while ((weakReferenceResults = referenceQueueSyncRealmResults.poll()) != null ) {
+            syncRealmResults.remove(weakReferenceResults);
+        }
+        while ((weakReferenceObject = referenceQueueRealmObject.poll()) != null ) {
+            realmObjects.remove(weakReferenceObject);
+        }
+    }
+
+    void notifyAsyncRealmResultsCallbacks() {
+        notifyRealmResults(asyncRealmResults.keySet().iterator());
+    }
+
+    void notifySyncRealmResultsCallbacks() {
+        notifyRealmResults(syncRealmResults.keySet().iterator());
+    }
+
+    void notifyRealmObjectCallbacks() {
```

Comment:
Shouldn't this either be `notifyRealmObjects` or change `notifyRealmResults` to `notifyRealmResultCallbacks` for consistent naming?